### PR TITLE
Improvements & Fixes

### PR DESCRIPTION
**Pull request by Martinski4GitHub**
- Made improvements in the code that builds the cron schedule string for the automatic script updates to make it more robust.

- Made improvements in the cron schedule validation code & some minor wording changes in the error messages generated when validation fails.

- Fixed bugs when trying to determine if we have a "386" build version in order to select the corresponding changelog URL.

- Various code improvements & fine-tuning.

### CODE REVIEW

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

Whenever you get a chance, please review this PR. No major changes, but several minor changes and one moderate change were made, along with a couple of fixes. Your 2nd pair of eyes is always welcome to double-check :>).

BTW, the initial report from the Linter tool found 3 new warnings - nothing major. After making the changes this evening, I re-ran the latest script and it came out "clean" as a whistle. :>)

Have a good night, bud.

**Review by ExtremeFiretop — APPROVED** [review]

**Comment by ExtremeFiretop** [comment]
Well I just give everything a quick test and it seems all is working!
Let me know if your ready for the next release :)

**Comment by Martinski4GitHub** [comment]
> Well I just give everything a quick test and it seems all is working! Let me know if your ready for the next release :)
>
>
I have nothing else to add for this release, so you can issue the production version at any time you're ready.

BTW, earlier this afternoon I ordered from Amazon three new RT-AX86U-PRO models that were on sale for "Black Friday" ($179.99 USD). One for my parents, the other for my in-laws (currently both households have the RT-AC68U), and another one for my home. When I talked with both my parents & in-laws yesterday during our Thanksgiving celebration, they told me to look for a replacement <= $200, and the PRO model was just in the right price range so I decided to also get one for myself to give it a try. It looks like a really good replacement for the AC models we have and for the right price. We'll see how it goes...

**Comment by ExtremeFiretop** [comment]
> > Well I just give everything a quick test and it seems all is working! Let me know if your ready for the next release :)
> 
> I have nothing else to add for this release, so you can issue the production version at any time you're ready.
> 

Fantastic, I actually have a bit of time this evening so I may have a quick look at your recommendation about the changelog logic.

I know the fix we applied worked but I'm going to look through past PRs to see how it developed into the solution we currently have (many fixes over time) and get a sense of there's any reason we couldn't adjust the logic.

If I find anything or decide it's not a quick thing ill just issue a production releases as is.

> BTW, earlier this afternoon I ordered from Amazon three new RT-AX86U-PRO models that were on sale for "Black Friday" ($179.99 USD). One for my parents, the other for my in-laws (currently both households have the RT-AC68U), and another one for my home. When I talked with both my parents & in-laws yesterday during our Thanksgiving celebration, they told me to look for a replacement <= $200, and the PRO model was just in the right price range so I decided to also get one for myself to give it a try. It looks like a really good replacement for the AC models we have and for the right price. We'll see how it goes...

Fantastic choice! It officially is getting the 3006 updates, and is supported by Merlin, so with some luck, eventually Merlin will move it over to the 3006 codebase with the other "PRO" AX models!

Looking at the specs, it's a decent router for the price! Especially with a black Friday sale! I've been itching to buy something, probably a new motherboard or a new wireless headset for work, before the tarrifs come down anyways 😆😉

**Comment by Martinski4GitHub** [comment]
> > > Well I just give everything a quick test and it seems all is working! Let me know if your ready for the next release :)
> > 
> > 
> > I have nothing else to add for this release, so you can issue the production version at any time you're ready.
> 
> Fantastic, I actually have a bit of time this evening so I may have a quick look at your recommendation about the changelog logic.
> 
> I know the fix we applied worked but I'm going to look through past PRs to see how it developed into the solution we currently have (many fixes over time) and get a sense of there's any reason we couldn't adjust the logic.
> 
> If I find anything or decide it's not a quick thing ill just issue a production releases as is.
>
OK, sounds good. As you said, the current solution works but if you have the time before the production release to make it better and more robust, go for it!! :>).

> 
> > BTW, earlier this afternoon I ordered from Amazon three new RT-AX86U-PRO models that were on sale for "Black Friday" ($179.99 USD). One for my parents, the other for my in-laws (currently both households have the RT-AC68U), and another one for my home. When I talked with both my parents & in-laws yesterday during our Thanksgiving celebration, they told me to look for a replacement <= $200, and the PRO model was just in the right price range so I decided to also get one for myself to give it a try. It looks like a really good replacement for the AC models we have and for the right price. We'll see how it goes...
> 
> Fantastic choice! It officially is getting the 3006 updates, and is supported by Merlin, so with some luck, eventually Merlin will move it over to the 3006 codebase with the other "PRO" AX models!
> 
> Looking at the specs, it's a decent router for the price! Especially with a black Friday sale! I've been itching to buy something, probably a new motherboard or a new wireless headset for work, before the tarrif war starts anyways 😆😉
>
>
I think for the RT-AC68U models, this PRO model is a great replacement, especially given the use-case for my parents & in-laws, and it may even be a bit more than they really need for their home network. They don't have any WiFi7 devices, have no need for VLANs, and their total number of clients (wired & wireless) is about 20. In my case, I want to give it a try and check out the WiFi coverage, and I still have the option to add the RT-AC86U as an AP. We'll see... But I think I'll have some fun setting them up for my parents & in-laws.

**Comment by ExtremeFiretop** [comment]
> 
> I think for the RT-AC68U models, this PRO model is a great replacement, especially given the use-case for my parents & in-laws, and it may even be a bit more than they really need for their home network. They don't have any WiFi7 devices, have no need for VLANs, and their total number of clients (wired & wireless) is about 20. In my case, I want to give it a try and check out the WiFi coverage, and I still have the option to add the RT-AC86U as an AP. We'll see... But I think I'll have some fun setting them up for my parents & in-laws first.

I think you did the right call, it's probably a little over kill for a regular user like them, but for you it's going to be just right.

Plus like you said you can experiment with setting up the old one as a node and see how that works for expanding your coverage. 

Obviously the AC won't reach the speeds the AX router will, but for testing purposes and extra coverage it's should do the job without a pickup.

**Comment by Martinski4GitHub** [comment]
> > I think for the RT-AC68U models, this PRO model is a great replacement, especially given the use-case for my parents & in-laws, and it may even be a bit more than they really need for their home network. They don't have any WiFi7 devices, have no need for VLANs, and their total number of clients (wired & wireless) is about 20. In my case, I want to give it a try and check out the WiFi coverage, and I still have the option to add the RT-AC86U as an AP. We'll see... But I think I'll have some fun setting them up for my parents & in-laws first.
> 
> I think you did the right call, it's probably a little over kill for a regular user like them, but for you it's going to be just right.
> 
> Plus like you said you can experiment with setting up the old one as a node and see how that works for expanding your coverage.
> 
> Obviously the AC won't reach the speeds the AX router will, but for testing purposes and extra coverage it's should do the job without a hickup.
>
>
My primary concern is not so much speed but getting a strong & consistent signal all the way to the corners and far side of our patio, mostly for spring & summer afternoons/evenings when we're hanging out on the patio. It's not a critical requirement, but it would be nice if it works well having the AC86U as an AP (not an AiMesh node).

File: MerlinAU.sh [hunk 1]
```diff
@@ -706,21 +708,26 @@ else
...
+readonly CRON_DAYofMONTH_rexp3="${CRON_DAYofMONTH_rexp2}[/][2-9]"
+readonly CRON_DAYofMONTH_rexp4="${CRON_DAYofMONTH_rexp1}([,]${CRON_DAYofMONTH_rexp1})+"
+readonly CRON_DAYofMONTH_RegEx="($CRON_DAYofMONTH_rexp1|$CRON_DAYofMONTH_rexp2|$CRON_DAYofMONTH_rexp3|$CRON_DAYofMONTH_rexp4)"
 
-readonly CRON_DAYofWEEK_NAME="([S|s]un|[M|m]on|[T|t]ue|[W|w]ed|[T|t]hu|[F|f]ri|[S|s]at)"
-readonly CRON_DAYofWEEK_rexp1="${CRON_DAYofWEEK_NAME}|[0-6]"
-readonly CRON_DAYofWEEK_rexp2="${CRON_DAYofWEEK_NAME}[-]${CRON_DAYofWEEK_NAME}|[0-6][-][0-6]"
-readonly CRON_DAYofWEEK_rexp3="${CRON_DAYofWEEK_NAME}([,]${CRON_DAYofWEEK_NAME})+|[0-6]([,][0-6])+"
-readonly CRON_DAYofWEEK_RegEx="($CRON_DAYofWEEK_rexp1|$CRON_DAYofWEEK_rexp2|$CRON_DAYofWEEK_rexp3)"
+readonly CRON_DAYofWEEK_Names="([S|s]un|[M|m]on|[T|t]ue|[W|w]ed|[T|t]hu|[F|f]ri|[S|s]at)"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:725]: I just realized I left all my comments pending on this review again. Whoops my bad... Submitting review now!

Approved regex changes

**Comment by ExtremeFiretop** [comment]
> OK, sounds good. As you said, the current solution works but if you have the time before the production release to make it better and more robust, go for it!! :>).
> 

I found the PRs...
Specifically: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/107
and:
https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/198

Just trying to get more information as to what prompted those changes from me. Especially 198... Little detail in the PR and you didn't even get to review it; makes me think it must of been a reported bug but where in the threads... Lol

**Comment by ExtremeFiretop** [comment]
Okay; I now understand why we ended up with this solution.
Basically with the code up top; the requirements include (currently)

1. Drop the prefix for comparison with the changelog file
2. Drop any suffix zeros for comparison with the changelog file, but allow any other suffix NUMBER
3. Drop any suffix that includes non-numerical values, examples below:
Suffix: "0_rog" should become "0" 
Suffix: "2_rog" should become "2"
Suffix: "beta3" should be dropped completely

Now the idea moving forwards would be to move that logic out of the top code and instead try to incorporate it into the Regex.
For now it works; and my brain isn't fully there to think about replacement just yet. But I've more or less understood how we ended up here and the requirements that need to be factored into the new solution whatever it may be.

**Comment by Martinski4GitHub** [comment]
>
> Okay; I now understand why we ended up with this solution. Basically with the code up top; the requirements include (currently)
> 
>     1. Drop the prefix for comparison with the changelog file (in new solution this can be building specific)
> 
>     2. Drop any suffix zeros (.0 or _0) for comparison with the changelog file, but allow any other suffix NUMBER (.2 or _2)
> 
>     3. Drop any suffix that includes non-numerical values, examples below:
>        Suffix: "0_rog" should become "0"
>        Suffix: "2_rog" should become "2"
>        Suffix: "beta3" should be dropped completely
> 
> 
> Now the idea moving forwards would be to move that logic out of the top code and instead try to incorporate it into the Regex. For now it works; and my brain isn't fully there to think about replacement just yet. But I've more or less understood how we ended up here and the requirements that need to be factored into the new solution whatever it may be.
>
>
OK, this is good. Making a list of requirements like that in such detail is an excellent way to analyze the problem and figure out a more robust & elegant solution.

I see that you got the production release out in the field. Good job, bud. Hopefully, things will be calm (i.e. no major bugs) for the rest of the year so we get more rest & more free time during the holidays. LOL!!!

Take care, bud, and have a great weekend.

**Comment by ExtremeFiretop** [comment]
> Hopefully, things will be calm (i.e. no major bugs) for the rest of the year so we get more rest & more free time during the holidays. LOL!!!

Agreed! That's my plan, over the next few weeks I may or may not be available as we get closer to the holidays, my plan is to focus on the upcoming Christmas and new year's plans and celebrations. I wanted to get everything out the door so we could take the next month to focus on other things (unless some unexpected bug comes up) but i know you've been busy as well. I think we did great work this release, lots of good stuff done.

> 
> Take care, bud, and have a great weekend.

You as well! Chat soon.

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

Just an FYI that we did get a bug report here:

https://www.snbforums.com/threads/merlinau-v1-3-7-the-ultimate-firmware-auto-updater-gnuton-support.91326/post-936071

I can't replicate the issue but I'm assuming based on his debug messages that it's coming from "dd" in this function:

##----------------------------------------##
## Modified by Martinski W. [2024-Jul-30] ##
##----------------------------------------##
_GetRawKeypress_() 
{
   local savedSettings
   savedSettings="$(stty -g)"
   stty -echo raw
   echo "$(dd count=1 2>/dev/null)"
   stty "$savedSettings"
}

Google tells me:

'dd typically also requires a block size (bs) to function predictably.'

**Comment by Martinski4GitHub** [comment]
> @Martinski4GitHub
> 
> Just an FYI that we did get a bug report here:
> 
> https://www.snbforums.com/threads/merlinau-v1-3-7-the-ultimate-firmware-auto-updater-gnuton-support.91326/post-936071
> 
> I can't replicate the issue but I'm assuming based on his debug messages that it's coming from "dd" in this function:
> 
> ```
> ##----------------------------------------##
> ## Modified by Martinski W. [2024-Jul-30] ##
> ##----------------------------------------##
> _GetRawKeypress_() 
> {
>    local savedSettings
>    savedSettings="$(stty -g)"
>    stty -echo raw
>    echo "$(dd count=1 2>/dev/null)"
>    stty "$savedSettings"
> }
> ```
> 
> Google tells me:
> 
> 'dd typically also requires a block size (bs) to function predictably.'
>
Yeah, I saw the post ~8 minutes ago, and the real issue is the current **PATH** set up by the user's environment.
I just submitted a PR to address the problem.

**Pull request merged**